azkaban-common/src/main/java/azkaban/server/session/SessionCache.java

```diff
@@ -75,11 +92,42 @@ public long getSessionCount() {
     return this.cache.size();
   }
 
+
+  /**
+   * @return <code>true</code> The number of cached session sharing the same IP and user equals
+   * or greater than the allowed limit if defined;
+   * <code>false</code> otherwise.
+   */
+  private boolean isViolatingMaxNumberOfSessionPerIpPerUser(final Session session) {
+    if (this.maxNumberOfSessionsPerIpPerUser.isPresent()) {
+      // get sessions sharing the same IP and user
+      final Set<Session> sessionsWithSameIP = this.findSessionsByIP(session.getIp());
+      int duplicateSessionCount = 0;
+      for (final Session sessionByIP : sessionsWithSameIP) {
```

Comment:
Talked offline about encapsulating this logic in a new method findSessionsByIPByUser or similar name.

Reply:
added more comments to make this piece more understandable.

Reply:
updated: extract this piece to a new function.